components/esmterrain/storage.cpp

```diff
@@ -385,72 +380,61 @@ namespace ESMTerrain
-            textureIndicesMap[*it] = size;
-            layerList.push_back(getLayerInfo(getTextureName(*it)));
-        }
-
-        // size-1 since the base layer doesn't need blending
-        int numBlendmaps = textureIndices.size() - 1;
 
-        // Second iteration - create and fill in the blend maps
         const int blendmapSize = (realTextureSize-1) * chunkSize + 1;
         // We need to upscale the blendmap 2x with nearest neighbor sampling to look like Vanilla
         const int imageScaleFactor = 2;
         const int blendmapImageSize = blendmapSize * imageScaleFactor;
 
-        for (int i=0; i<numBlendmaps; ++i)
-        {
-            osg::ref_ptr<osg::Image> image (new osg::Image);
-            image->allocateImage(blendmapImageSize, blendmapImageSize, 1, GL_ALPHA, GL_UNSIGNED_BYTE);
-            unsigned char* pData = image->data();
+        LandCache cache;
+        std::map<UniqueTextureId, unsigned int> textureIndicesMap;
 
-            for (int y=0; y<blendmapSize; ++y)
+        for (int y=0; y<blendmapSize; ++y)
+        {
+            for (int x=0; x<blendmapSize; ++x)
             {
-                for (int x=0; x<blendmapSize; ++x)
+                UniqueTextureId id = getVtexIndexAt(cellX, cellY, x+rowStart, y+colStart, cache);
+                std::map<UniqueTextureId, unsigned int>::iterator found = textureIndicesMap.find(id);
+                if (found == textureIndicesMap.end())
                 {
-                    UniqueTextureId id = getVtexIndexAt(cellX, cellY, x+rowStart, y+colStart, cache);
-                    assert(textureIndicesMap.find(id) != textureIndicesMap.end());
-                    int layerIndex = textureIndicesMap.find(id)->second;
+                    unsigned int layerIndex = layerList.size();
+                    Terrain::LayerInfo info = getLayerInfo(getTextureName(id));
 
-                    int alpha = (layerIndex == i+1) ? 255 : 0;
+                    // look for existing diffuse map, which may be present when several plugins use the same texture
+                    for (unsigned int i=0; i<layerList.size(); ++i)
+                    {
+                        if (layerList[i].mDiffuseMap == info.mDiffuseMap)
+                        {
+                            layerIndex = i;
+                            break;
+                        }
+                    }
 
-                    int realY = (blendmapSize - y - 1)*imageScaleFactor;
-                    int realX = x*imageScaleFactor;
+                    found = textureIndicesMap.insert(std::make_pair(id, layerIndex)).first;
 
-                    pData[(realY+0)*blendmapImageSize + realX + 0] = alpha;
-                    pData[(realY+1)*blendmapImageSize + realX + 0] = alpha;
-                    pData[(realY+0)*blendmapImageSize + realX + 1] = alpha;
-                    pData[(realY+1)*blendmapImageSize + realX + 1] = alpha;
+                    if (layerIndex >= layerList.size())
```

Comment:
layerIndex will never be greater than the size of the list. In fact, the whole layer index handling in this scope is obscure. Not sure how exactly to adjust the surrounding code with that in mind, though.

Reply:
But it still can be **equal** to the size of list. It seems the > is here just for sure.